src/extension.ts

```diff
@@ -0,0 +1,243 @@
+
+const opcodeNames = [
+  "ld",
+  "st",
+  "mov",
+  "add",
+  "and",
+  "inc",
+  "inca",
+  "dec",
+  "deca",
+  "not",
+  "gpc",
+  "shr",
+  "shl",
+  "br",
+  "beq",
+  "bgt",
+  "j",
+  "halt",
+  "nop",
+] as const;
+
+type OpcodeName = typeof opcodeNames[number];
+function isOpcodeName(x: string): x is OpcodeName {
+  // widen formats to string[] so indexOf(x) works
+  return (opcodeNames as readonly string[]).indexOf(x) >= 0;
+}
+
+/**
+ * Signature of an instruction
+ *
+ * containing information of input argument(s)
+ */
+type Signature = {
+  arg1?: Argument;
+  arg2?: Argument;
+};
+
+/**
+ * Class representing an input argument of an SM213 instruction
+ */
+class Argument {
+  /** name of the argument */
+  argName: string;
+  addressingModes: AddressingMode[];
+  /** description of the argument
+   *
+   * Format: <argName: addressingMode1/addressingMode2...>
+   */
+  desc: string;
+  constructor(argName: string, modes: AddressingMode[]) {
+    this.argName = argName;
+    this.addressingModes = modes;
+    // Generate desc
+    // in the format of <argName: mode1/mode2...>
+    this.desc = `<${this.argName}: `;
+    for (const mode of modes) {
+      this.desc += `${mode}/`;
+    }
+    this.desc = this.desc.replace(/.$/, ">");
+  }
+}
+
+/**
+ * Class representing an Opcode
+ */
+class Opcode {
+  name: OpcodeName;
+  signature?: Signature;
+  /**
+   * Description of the Opcode
+   *
+   * It will be displayed as the second part of the hover preview
+   */
+  description: string;
+  markdownDesc: vscode.MarkdownString;
+
+  /**
+   *
+   * @param description description of the Opcode, for which $1, $2 are used as placeholder for argument names
+   */
+  constructor(name: OpcodeName, description: string, signature?: Signature) {
+    this.name = name;
+    this.signature = signature;
+    this.description = description;
+    // Format description from signature
+    if (this.signature?.arg1) {
+      this.description = this.description.replace(
+        "$1",
+        String.raw`\<${this.signature.arg1.argName}\>`
+      );
+    }
+    if (this.signature?.arg2) {
+      this.description = this.description.replace(
+        "$2",
+        String.raw`\<${this.signature.arg2.argName}\>`
+      );
+    }
+
+    // Generate markdownDesc
+    let heading = this.name;
+    if (this.signature?.arg1) {
+      heading += ` ${this.signature?.arg1.desc}`;
+    }
+    if (this.signature?.arg2) {
+      heading += ` ${this.signature?.arg2.desc}`;
+    }
+    this.markdownDesc = new vscode.MarkdownString()
+      .appendCodeblock(heading, "asm")
+      .appendMarkdown("\n --- \n") // insert horizontal rule
+      .appendMarkdown(this.description);
+  }
+}
+
+const OPCODES = new Map<OpcodeName, Opcode>([
+  [
+    "ld",
+    new Opcode("ld", "loads value from $1 to $2", {
+      arg1: new Argument("val", ["immediate", "base+offset", "indexed"]),
+      arg2: new Argument("dest", ["register"]),
+    }),
+  ],
```

Comment:
> Minor nit: I don't know what the best way to do this is, but I feel like it would add to the overall clarity if we could distinguish between immediate addressing (which requires an actual value or a label) and base+offset and indexed (which needs a register). I understand that it's hard to fit all that info in the hover box nicely, so I'm good with this as is now too

I want to make sure that I am understanding it clearly. So I assume that the current hovering header is not good enough?

> Another thing I don't know how to solve: when I took 213, I found it hard to tell whether an instruction was reading from memory or not. I'm wondering if it would be possible to add a little something about determining which register in an instruction can access memory (maybe this could just be more info in the addressing mode description part of the hover?)

I believe that this is a fantastic little addition, although I am unsure about how exactly it should be done within the already clustered addressing mode description line (the heading line). My current idea is that addition of semantic & improved word description would make it clear.

Reply:
> I want to make sure that I am understanding it clearly. So I assume the suggestion is that base+offset and indexed should be in some way, grouped together to enhance clarity?

I think so, just to sort of hammer down that addressing modes (both base+offset and indexed) need a register whose value is an address whereas immediates are a label or a `$v` where `v` is a number in decimal or hex

> I believe that this is a fantastic little addition, although I am unsure about how exactly it should be done within the already clustered addressing mode description line (the heading line). My current idea is that (future) addition of semantic & improved word description are sufficient to make it clear, but I welcome further critical comments.

For sure! Again, just spitballing some rough ideas here, so these could be explored further (if we want to) later on